server/src/test/java/org/opensearch/action/pagination/ShardPaginationStrategyTests.java

```diff
@@ -0,0 +1,400 @@
+        List<Integer> indexNumberList = new ArrayList<>();
+        List<Integer> deletedIndices = new ArrayList<>();
+        final int totalIndices = 100;
+        final int numIndicesToDelete = 10;
+        final int numIndicesToCreate = 5;
+        final int fixedIndexNumToDelete = 7;
+        final int numShardsForNewIndices = 4;
+        final int numReplicasForNewIndices = 5;
+        final int totalInitialShards = totalIndices * DEFAULT_NUMBER_OF_SHARDS * (DEFAULT_NUMBER_OF_REPLICAS + 1);
+        for (int indexNumber = 1; indexNumber <= 100; indexNumber++) {
+            indexNumberList.add(indexNumber);
+        }
+        ClusterState clusterState = getRandomClusterState(indexNumberList);
+
+        int pageSize = 10;
+        String requestedToken = null;
+        int numPages = 0;
+        List<ShardRouting> shardRoutings = new ArrayList<>();
+        Set<String> indicesFetched = new HashSet<>();
+        do {
+            numPages++;
+            PageParams pageParams = new PageParams(requestedToken, PARAM_ASC_SORT_VALUE, pageSize);
+            ShardPaginationStrategy paginationStrategy = new ShardPaginationStrategy(pageParams, clusterState);
+            assertNotNull(paginationStrategy);
+            assertNotNull(paginationStrategy.getResponseToken());
+            requestedToken = paginationStrategy.getResponseToken().getNextToken();
+            // deleting test-index-7 & 10 more random indices after 11th call. By that time, shards for first 6
+            // indices would have been completely fetched. 11th call, would fetch first 3 shardsIDs for test-index-7 and
+            // if it gets deleted, rest 2 shardIDs for it should not appear.
+            if (numPages == 11) {
+                clusterState = deleteIndexFromClusterState(clusterState, fixedIndexNumToDelete);
+
+                deletedIndices = indexNumberList.subList(20, indexNumberList.size());
+                Collections.shuffle(deletedIndices, getRandom());
+                for (int pos = 0; pos < numIndicesToDelete; pos++) {
+                    clusterState = deleteIndexFromClusterState(clusterState, deletedIndices.get(pos));
+                }
+            }
+            // creating 5 indices after 5th call
+            if (numPages == 5) {
+                for (int indexNumber = totalIndices + 1; indexNumber <= totalIndices + numIndicesToCreate; indexNumber++) {
+                    clusterState = addIndexToClusterState(clusterState, indexNumber, numShardsForNewIndices, numReplicasForNewIndices);
+                }
+            }
+            assertTrue(paginationStrategy.getRequestedEntities().size() <= pageSize);
+            shardRoutings.addAll(paginationStrategy.getRequestedEntities());
+            indicesFetched.addAll(paginationStrategy.getRequestedIndices());
+        } while (Objects.nonNull(requestedToken));
+
+        assertEquals(totalIndices + numIndicesToCreate - numIndicesToDelete, indicesFetched.size());
```

Comment:
- Can this number be fetched from the clusterState ?
- should we iterate over the clusterState and verify indices and shardRoutings ?